feat(ecmascript): `String.prototype.split()`

oscarotero opened the pull request:
Please, bear with me. I'm learning Rust and I don't have much experience with the borrow checker.

oscarotero commented:
@aapoalas I think it's ready to review.

btw, what is the recommended way to test the function? I'd like to have a file to execute (like `cargo run -- eval testing.js`) to manually test the function, but there isn't a `console.log()` that I can use to see the result.

aapoalas commented:
`nova_cli` has a `print` function you can use.

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -732,8 +733,73 @@ impl StringPrototype {
+        // 5. Let separatorStr be ? ToString(separatorString).
+        let separator_value = args.get(0);
+
+        // 6. If separator is undefined, return an array with this.
```
**issue (nonblocking):** This should be spec step 7, and this is also not the spec text.

I won't block this PR on this, but in the future please leave the spec text in the code mostly as-is. Commenting on the spec text with eg. "Note: ..." is fine. Moving spec text and steps around is also fine as long as it doesn't change any observable features of the function. Oftentimes leaving a note about these sort of reorderings is a good idea.

But rewriting the spec text in your own words, and then in the process having it turn into just your custom algorithm with steps outlined is not a good thing. If someone else comes to look at this function and tries to figure out how this corresponds to the spec algorithm, they'll have very little idea what is going on. Maybe this is written based on some old spec text that is no longer available online? And in that case, what has changed? Presumably the algorithm has been changed in meaningful ways and then this old code written to the old spec is no longer valid; but how?

It's better to abide by the spec text as much as possible, and only make additions (fast paths) and minor reorderings where it makes sense. That way a future reader (which may be you yourself) has an easier time understanding what is being done and why, and if the spec changes they'll also know what they need to change to fix the code.

oscarotero replied:
Sorry, I didn't realize these comments were spec comments. I though they were simple comments to follow easier what the code does.
I'll take a look to that.

aapoalas commented:
Oh, oops. I only now realized this is actually your first PR here! Sorry, I wasn't paying attention and thought you were behind #416 as well!

Welcome, and great work getting yourself up from the ground with the borrow checker! And sorry about the nitpicking on the spec text: It's just a good idea to stick fairly close to it. I don't mean it personally <3

oscarotero commented:
Thanks for your review @aapoalas 
Writing a JS engine is harder than I thought :D

I'm learning the codebase of the repo, how to implement and test things, and now I just learned that the numbers in the comments corresponds to the steps in the ecma262 specs.

I'll continue working on this.

aapoalas commented:
> I'm learning the codebase of the repo, how to implement and test things, and now I just learned that the numbers in the comments corresponds to the steps in the ecma262 specs.

I'm sorry! This is definitely my failing! I should add some more about this to the CONTRIBUTING.md and maybe elsewhere as well so that new contributors have an easier way in.

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
+        let limit = args.get(0);
+        let lim = match limit {
+            Value::Undefined => u32::MAX,
+            _ => to_uint32(agent, limit)?,
```
**suggestion:** Add a `Value::Integer(value) => value.into_i64() as u32`

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
-        }
+        // 4. If limit is undefined, lim is u32::MAX - 1.
+        let limit = args.get(0);
+        let lim = match limit {
```
**praise:** I've not actually thought about combining spec `if undefined { X } else { Y }` into matches like this. Makes for fairly pretty code and makes it possible to add fast paths pretty trivially. Nice! <3

oscarotero replied:
happy to know at least I did something well! :D

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
+        // 4. If limit is undefined, lim is u32::MAX - 1.
+        let limit = args.get(0);
+        let lim = match limit {
+            Value::Undefined => u32::MAX,
```
**issue:** Should be `u32::MAX - 1`

oscarotero replied:
The spec says `2**32 - 1`. I don't know if `u32::MAX` is that number or it should substract 1.
(I know the comment is wrong)

oscarotero replied:
okay, fixed!

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
 
-            return Ok(results.into_value());
-        }
+        // 4. If limit is undefined, lim is u32::MAX - 1.
```
**nitpick:** You're still paraphrasing the spec. I'd just copy it verbatim:

```rs
let lim = match limit {
  // 4. If limit is undefined, let lim be 2**32 - 1;
  Value::Undefined => u32::MAX - 1,
  // else let lim be ℝ(? ToUint32(limit)).
  // Note: Fast path for integer parameter.
  Value::Integer(limit) => limit.into_i64() as u32,
  _ => to_uint32(agent, limit)?,
};
```

oscarotero replied:
fixed

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
-            let a = array_create(agent, 0, 0, None)?;
-            return Ok(a.into_value());
+        // 7. If separator is undefined, return an array with the whole string
+        if let Value::Undefined = separator {
```
**nitpick:** Could just use `separator.is_undefined()` here.

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
-            return Ok(a.into_value());
+        // 7. If separator is undefined, return an array with the whole string
+        if let Value::Undefined = separator {
+            let list: [Value; 1] = [s.into_value()];
```
**nitpick:** You can just do `return Ok(create_array_from_list(agent, &[s.into_value()]);` to avoid having to define the array type.

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
-        };
+        // 9. If separatorLength = 0, the split by characters
+        if separator_length == 0 {
+            let subject = s.as_str(agent).to_owned();
```
**nitpick:** I would leave in the spec steps a through e, BUT also **praise:** Explaining that this is doing a split by characters is very nice!

**suggestion:** I _think_ we're guaranteed that all of the characters are necessarily expressible within 7 bytes as UTF-8. I think so anyway. As such, I think you can actually avoid the `to_owned()` here by just asserting that each split part will pass `SmallString::try_from(part).unwrap()`.

There would also be a way to avoid the Vec altogether and just shuffle data within the heap directly, but we're missing a `ArrayHeapMutIndexable` struct for now. `map_constructor.rs:172` and the `ArrayHeapIndexable` struct may serve as hints as to what I'm thinking about here. Just add in the knowledge that any UTF-16 character (which is what we should be splitting the string into) should fit into 7 bytes when converted into UTF-8, I think, and all should be "clear" :D

oscarotero replied:
I've refactored this part.

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
+        // 9. If separatorLength = 0, the split by characters
+        if separator_length == 0 {
+            let subject = s.as_str(agent).to_owned();
+            let head = subject.split("");
```
**though:** I wonder if this does exactly the correct thing: The spec deals with UTF-16, but we're in UTF-8 world here. That makes things complicated sometimes. @andreubotella might understand this more deeply.

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
-        if separator.is_empty() {
+            // Note: Rust's split inserts an empty string in the beginning and end of the array. We remove them
             if results.len() > 1 {
                 results.remove(0);
```
**suggestion:** It would be better to do this "filtering" out as part of the enumeration itself.

You can probably do something like: `let results: Vec<Value> = head.enumerate().skip(1).take_while(lim).map(|part| Value::from_str(agent, part)).collect();`. Now when creating this from slice, just drop the last item if the length of the Vec is longer than 0.

oscarotero replied:
done

aapoalas reviewed (approved):
Found a bunch to comment on, but nothing terribly out of whack. Thank you very much, and welcome again to Nova!

Absolutely feel free to keep on trying to solve any remaining cases: The `eval-test` command is very useful for testing individual test262 case. eg.
```sh
cargo build && cargo run --bin test262 eval-test built-ins/String/prototype/split/argument-is-new-reg-exp-and-instance-is-string-hello.js
```

oscarotero commented:
@aapoalas I have fixed a lot of tests due a stupid bug: I was getting the limit argument with `args.get(0)` instead of `args.get(1)`. The only failing tests are those related with regexp, `Symbol.split` and other advanced edge cases.

I've also applied your suggestions. The only unresolved comment is the one related with UTF-16.

oscarotero commented:
No problem! I saw the pictures and completely understand your priorities! 🍇

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -732,8 +734,87 @@ impl StringPrototype {
+        let separator = args.get(0);
+
+        if matches!(separator, Value::Undefined | Value::Null) {
+            // TODO: get_method(separator, Symbol.split)
```
**issue:** This shouldn't be too hard: `get_method` function already exists, and then you just perform a `call_function(splitter, separator, Some(&[o, limit]))`

oscarotero replied:
I didn't implement this because I don't know how to create the `Symbol.split` value in Rust.
But if you help me with that, I can implement this.

aapoalas replied:
Ah, these "well-known symbols" can be "created" as simply as `WellKnownSymbolIndexes::Split.into()` :)

oscarotero replied:
Done in the last commit.

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -732,8 +734,87 @@ impl StringPrototype {
+                .enumerate()
+                .skip(1) // Rust's split inserts an empty string in the beginning.
+                .take_while(|(i, _)| *i < lim as usize)
+                .map(|(_, part)| Value::from(SmallString::try_from(part).unwrap()))
```
**nitpick:** Prefer `.into_value()` over `Value::from()`.

aapoalas commented on nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs:
```diff
@@ -732,8 +734,87 @@ impl StringPrototype {
+                .collect();
+
+            // Remove the latest empty string if exists
+            results.pop();
```
**question:** Are you sure this works currectly? What happens if limit is less than the length of the iterator? Won't this then drop one string that was supposed to be received?

oscarotero replied:
This case only happens when the separator is an empty string. You can see the documentation of `split()`:

```rs
/// When the empty string is used as a separator, it separates
/// every character in the string, along with the beginning
/// and end of the string.

let f: Vec<_> = "rust".split("").collect();
assert_eq!(f, &["", "r", "u", "s", "t", ""]);
```

In JavaScript `"rust".split("")` must return `["r", "u", "s", "t"]`, so that's why I remove the first and last elements. The first is removed in  `skip(1)`. The last is removed with `results.pop()`, that, according to the docs:
```
Removes the last element from a vector and returns it, or [`None`] if it is empty.
```

I think this code should work fine (all related tests are passed), but I'm open to change it if you want to implement it differently.

aapoalas replied:
Yeah, but now what happens if we do : `"rust".split("", 2)`. This should return `["r", "u"]` but I _think_ it will currently return just `["r"]` because we drop the last one regardless of the limit.

oscarotero replied:
That's a good point. And, indeed, it doesn't work. I'm fixing it.

oscarotero replied:
Ok, I think it's now fixed:
- The `take_while` was wrong because it didn't take into account the skip() done before.
- Only removes the lastest string if the limit is bigger than the result length.

It's weird that there isn't any test for this use case. I would have detected earlier.

aapoalas reviewed (approved):
Minor nitpicks / thoughts / questions.

But looks good to me! Thank you for this, great work!

aapoalas commented:
Thank you again and great work! <3
